web3/providers/rpc.py

```diff
@@ -57,6 +74,94 @@ def make_request(self, method, params):
         return response_body
 
 
+class KeepAliveRPCProvider(BaseProvider):
+    """RPC-provider that handles HTTP keep-alive connection correctly.
+
+    HTTP client and underlying TCP/IP network connection is recycled across requests.
+    :class:`HTTPClient` connection pooling is used for connections.
+
+    Preferably create only one instance of KeepAliveProvider per process,
+    though the class has internal in-process cache for clients.
+    """
+
+    #: In-process client cache keyed by host:port -> HTTPClient
+    clients = {}
+
+    def __init__(self,
+                 host="127.0.0.1",
+                 port=8545,
+                 path="/",
+                 ssl=False,
+                 connection_timeout=10,
+                 network_timeout=10,
+                 concurrency=10,
+                 *args,
+                 **kwargs):
+        """Create a new RPC client with keep-alive connection pool.
+
+        :param concurrency: See :class:`geventhttpclient.HTTPClient`
+
+        :param connection_timeout: See :class:`geventhttpclient.HTTPClient`
+
+        :param network_timeout: See :class:`geventhttpclient.HTTPClient`
+        """
+        self.host = host
+        self.port = int(port)
+        self.path = path
+        self.ssl = ssl
+        self.connection_timeout = connection_timeout
+        self.network_timeout = network_timeout
+        self.concurrency = concurrency
+
+        super(KeepAliveRPCProvider, self).__init__(*args, **kwargs)
+
+        self.client = self.get_or_create_client()
+
+    def get_or_create_client(self):
+        from web3 import __version__ as web3_version
+
+        key = "{}:{}".format(self.host, self.port)
+        client = KeepAliveRPCProvider.clients.get(key)
+        if client:
```

Comment:
Here is a pattern that I've used a lot in the past and liked for doing caching like this

```python

def my_cached_method(self):
    if key not in cache:
        ... do things
        cache[key] = thing
    return cache[key]
```

I've found the control flow is easier to follow.  Just a suggestion.